src/cli/download_handler.rs

```diff
@@ -84,15 +97,65 @@ impl DownloadHandler {
     }
 
     pub fn run(&self) -> HandlerResult {
+        let destination = self.selection_destination_for_install()?;
+        self.destination_may_not_dir(&destination)?;
         let github = GithubClient::from_environment();
         let release = self.fetch_release(&github)?;
         let selected_asset = self.select_asset(release)?;
+        self.asset_may_not_be_archive(&selected_asset.name)?;
         let output_path = self.choose_output_path(&selected_asset.name);
         Self::download_asset(&github, &selected_asset, &output_path)?;
-        self.maybe_install(&selected_asset.name, &output_path)?;
+        self.maybe_install(&selected_asset.name, &output_path, destination)?;
         Ok(())
     }
 
+    fn selection_destination_for_install(&self) -> Result<Destination, HandlerError> {
+        let cwd = Self::cwd()?;
+        match self.install {
+            Install::No => Ok(Destination::Directory(cwd)),
+            Install::Yes(_) => match self.output.as_ref() {
+                Some(output) if output.is_dir() => Ok(Destination::Directory(output.clone())),
+                Some(output) => Ok(Destination::File(output.clone())),
+                None => Ok(Destination::Directory(cwd)),
+            },
+        }
+    }
+
+    fn destination_may_not_dir(&self, destination: &Destination) -> Result<(), HandlerError> {
+        if self.install.is_more_than_one() {
+            match destination {
+                Destination::File(x) => Err(HandlerError::new(format!(
+                    "Multiple install target (-I,--install-file) are selected \
+                        but output (-o,--output) is not a directory\n \
+                        Output: {:?}",
+                    x
+                ))),
+                Destination::Directory(_) => Ok(()),
+            }
+        } else {
+            Ok(())
+        }
```

Comment:
Invert the `if` condition and remove the `else` branch to have less indentation and be more clear